### ISSUE TITLE
Self hosting Excalidraw - Umbrella issue

### ISSUE THREAD

**Issue by kbariotis** [excalidraw/excalidraw#1772]
So I'm opening the self-hosting thread.

## Current architecture
First an architecture overview. Excalidraw is made of three main parts at the moment.

- The client, what you see when you visit `excalidraw.com` which is this repo
- The sockets server, which powers the collaboration feature, [it's this repo](https://github.com/excalidraw/excalidraw-room), a Node.js server running `socket.io`
- The store server, which powers the sharing feature, [it's this repo](https://github.com/excalidraw/excalidraw-json), a Python server running on GAE with some kind of a DB

At the moment, we are planning on publishing a Docker image of the Excalidraw client only. That means that people will be able to self-host the client but they won't be able to use sharing/collaboration features because the client will still have to talk to our cloud-hosted backend servers on which there are CORS restrictions.

## For now

For the image, we are planning on publishing soon (in the next few days), we have already stripped out the google tag manager but I see the visible sharing/collaboration icons as an issue. People will be frustrated by not understanding why they don't work.

I suggest we strip them out too on the Docker image only. Temporarily until we have a plan for the rest two services. It's the best UX we can offer IMO.

## For the future

Moving forward we would like people to be able to self-host a full-fledged Excalidraw environment. My suggestion is to publish a Docker image for each of the two backend servers so people can self-host these as well. Having separate images promotes better scalability and also follows the one image/per container practice. They will be able to maintain and also people can opt-out of a feature if they don't want to support, e.g. they only want collaboration features but not the storing feature.

Then we will be able to point the client image to specific URLs and it will simply work. 

For that to work, the socket server will be much simpler as it has no dependencies. On the other hand, the storing server has a DB dependency which we will have to consider. Ideally, it should be something easy to use and maintain, like a postgres or a redis instance. 

We could consider how the future of that service looks like, but for now we can focus on what it actually needs. 

Let me know of your thoughts and we can start planning right away.

**Comment by kbariotis** [comment]
Worth mentioning that atm:

- socket server [allows all domains](http://localhost:3000/#json=6323886937341952,otHeakoayx2yp5AEqxQ-6w)
- storing server has an [allowlist of domains](https://github.com/excalidraw/excalidraw-json/blob/master/main/main.py#L36)

**Comment by kjellkvinge** [comment]
just to pitch in an alternative approach. I have experimented with this, and created a bundle - single binary - which will work for selfhosting (on mac/win/linux/rasperry pi) on local network. (currently only link sharing is implemented)

https://github.com/kjellkvinge/excalidrawserver

Is this interesting? It was a fun hack :)

regards

**Comment by NMinhNguyen** [comment]
Made a comment in https://github.com/excalidraw/excalidraw-json/issues/76#issuecomment-650463667 but will post here for visibility:

> Also, just wanted to say that during our company-wide hackathon yesterday and today, I managed to self-host both the socket and storage server. The socket server basically required no changes. The storage server I decided to re-implement from scratch using Next.js (I'm more familiar with JavaScript than Python and there's lots of GAE dependencies, so a rewrite was a lot simpler) and S3 for blob storage. I'll be looking to open source this soon, so stay tuned :)

**Comment by kbariotis** [comment]
Thank you @NMinhNguyen for sharing. Indeed the socket server needs only a Dockerfile and it would be ready to be self hosted along the main Excalidraw client. 

For the storage one, we will have to think the current storage solution and storing diagrams in the filesystem makes sense I think. So with a Dockerfile and shared volume on the host, people will be able to self host that as well.

Thank you again

**Comment by kjellkvinge** [comment]
@NMinhNguyen did you have any issues when serving the site over HTTP? Or do you plan he use HTTPS in all cases?
For simplicity I wanted to be able to selfhost on local network, but then I had to disable the clientside crypto.

**Comment by NMinhNguyen** [comment]
> Thank you @NMinhNguyen for sharing. Indeed the socket server needs only a Dockerfile and it would be ready to be self hosted along the main Excalidraw client.

I don't actually have much Docker experience but we have a [PaaS solution](https://docs.cloudfoundry.org/buildpacks/node/index.html) (somewhat similar to GAE in some ways) that understands Node.js so all I had to do was push the source code + dependencies but yeah :)

I'll need to think about the best way to parameterise endpoint URLs in the `excalidraw` client. What I ended up doing was extracting the URLs into like `REACT_APP_BACKEND_V2_POST_URL` env vars store in [`.env`](https://create-react-app.dev/docs/adding-custom-environment-variables/#adding-development-environment-variables-in-env) and then running `yarn build:app:docker`. But you see, that requires a build :)

If you have any suggestions (or links to good Docker resources), I'm willing to look into those too.

> @NMinhNguyen did you have any issues when serving the site over HTTP? Or do you plan he use HTTPS in all cases?
> For simplicity I wanted to be able to selfhost on local network, but then I had to disable the clientside crypto.

I actually just used HTTPS so I didn't have any issues but you're right in that `window.crypto` is unavailable in insecure contexts https://stackoverflow.com/a/46468377

**Comment by NMinhNguyen** [comment]
So I've been thinking a bit more about this and I'm wondering if the easiest way to make the backend URLs configurable, is to have some kinda `@excalidraw/cli` package that you would invoke as follows:

```sh
REACT_APP_BACKEND_V2_GET_URL=... npx @excalidraw/cli build
# this then creates a `public` folder or something
```

The CLI itself could take flags like `--gitRef` for specifying the commit hash/Git tag or branch or something, and default to the main branch in this repo. Such a CLI, in my opinion, has the most flexibility in terms of hosting:

- you can take `public` folder and host on Vercel
- you can still have `Dockerfile` that uses the CLI to build the site and uses nginx to serve it

Let me know if you think this approach is wack 😛

**Comment by Turakar** [comment]
@NMinhNguyen, from my observations, the by far most common way of configuration for docker images is by environment variables. Sometimes there are also some command line parameters or config files involved, but not a manual build.

**Comment by NMinhNguyen** [comment]
> @NMinhNguyen, from my observations, the by far most common way of configuration for docker images is by environment variables. Sometimes there are also some command line parameters or config files involved, but not a manual build.

Essentially the problem here is that environment variables are read only once when you do `yarn build`. Unless you want to muck around with rewriting the build output, it’s pretty difficult to inject those variables. Further to that, most solutions I’ve seen will suffer from content hashing issues. A rebuild is simply the most robust way to inject those environment variables for a static website (CRA), but I’m open to suggestions.

**Comment by kjellkvinge** [comment]
in my experiment i can do:

```
REACT_APP_HOST="htps://my.server.com/" npm run build:app:docker
```

But this require that the backendurls use this variable. I solve this with a patch like this, which I apply in a build step:
```patch
diff --git a/src/data/index.ts b/src/data/index.ts
index f9c3bec..793d28e 100644
--- a/src/data/index.ts
+++ b/src/data/index.ts
@@ -24,12 +24,18 @@ export { loadFromBlob } from "./blob";
 export { saveAsJSON, loadFromJSON } from "./json";
 export { saveToLocalStorage } from "./localStorage";
 
-const BACKEND_GET = "https://json.excalidraw.com/api/v1/";
-
-const BACKEND_V2_POST = "https://json.excalidraw.com/api/v2/post/";
-const BACKEND_V2_GET = "https://json.excalidraw.com/api/v2/";
-
-export const SOCKET_SERVER = "https://excalidraw-socket.herokuapp.com";
+// get backend host from env. default excalidraw
+const HOST = process.env.REACT_APP_HOST
+  ? process.env.REACT_APP_HOST
+  : "https://json.excalidraw.com/";
+const BACKEND_GET = `${HOST}api/v1/`;
+
+const BACKEND_V2_POST = `${HOST}api/v2/post/`;
+const BACKEND_V2_GET = `${HOST}api/v2/`;
+
+export const SOCKET_SERVER = process.env.REACT_APP_SOCKET_SERVER
+  ? process.env.REACT_APP_SOCKET_SERVER
+  : "https://excalidraw-socket.herokuapp.com";
 
 export type EncryptedData = {
   data: ArrayBuffer;
```

**Comment by NMinhNguyen** [comment]
@kjellkvinge yup that is precisely what I did so I could rebuild it. But just wanted to agree on the approach first before making any changes, although I suppose it doesn't harm to introduce those environment variables and an `.env` file anyway: https://create-react-app.dev/docs/adding-custom-environment-variables/#adding-development-environment-variables-in-env

**Comment by Turakar** [comment]
> > @NMinhNguyen, from my observations, the by far most common way of configuration for docker images is by environment variables. Sometimes there are also some command line parameters or config files involved, but not a manual build.
> 
> Essentially the problem here is that environment variables are read only once when you do `yarn build`. Unless you want to muck around with rewriting the build output, it’s pretty difficult to inject those variables. Further to that, most solutions I’ve seen will suffer from content hashing issues. A rebuild is simply the most robust way to inject those environment variables for a static website (CRA), but I’m open to suggestions.

Ok, got your point. I was not aware of the limitations imposed by the build process and just had a look from the sysadmin perspective. Rebuilds are ugly to handle (in my opinion), but if there is no workaround :|

**Comment by kbariotis** [comment]
Thank you all for your ideas. I think a good small step would be to containerize the socket server. Please feel free to make a PR for that and we can then talk about integrating with the main app Docker image. :)

**Comment by NMinhNguyen** [comment]
@kbariotis I've raised a PR that makes it simpler to `git clone` this repo and `yarn build:app:docker` to produce build output with customised API endpoints. A Docker solution actually doesn't help my use case much, e.g. if I wanted to use something similar to Vercel to host my website - it only understands static files (or JS, Python etc.) and not Docker containers 😅

**Comment by NMinhNguyen** [comment]
From https://github.com/excalidraw/excalidraw-json/issues/76#issuecomment-652832953

> I've open sourced the Next.js storage server: https://github.com/NMinhNguyen/excalidraw-json. At the moment it only supports Amazon S3-compatible storage (e.g. [Google Cloud Storage](https://cloud.google.com/storage/docs/interoperability), [Yandex Object Storage](https://cloud.yandex.com/docs/storage/)) and only implements v2 of the API (i.e. the one with e2e encryption), but I might add an option to store files on the file system.

**Comment by bmullan** [comment]
Pardon my noob Excalidraw question but I just discovered Excalidraw today and have looked but is there _any guide to setup **self-hosted but without using Docker**_?

In my case I am a long time user of  LXD "system" containers and I'd like to host this in an LXD container.    

So even if all that is available is any non-formal text file describing how to install it on Debian, CentOS, Ubuntu, Fedora etc would be enough to install it in an LXD container and I'll give it a shot and provide feedback.

Thanks for any info anyone can provide.

Brian

**Comment by NMinhNguyen** [comment]
> Thank you all for your ideas. I think a good small step would be to containerize the socket server. Please feel free to make a PR for that and we can then talk about integrating with the main app Docker image. :)

@kbariotis I've raised https://github.com/excalidraw/excalidraw-room/pull/71 although I've left publishing out of the PR for now.

**Comment by NMinhNguyen** [comment]
By the way, I wanted to ask about scalability - how many instances of `excalidraw-json` and `excalidraw-room` do you have running for https://excalidraw.com/? I'm not too familiar with `excalidraw-room`, but from what I understand, you can only easily have 1 instance? Or you might need some sort of smart routing so that all requests with the same room ID get routed to the same https://socket.io/ server? At the same time, presumably even a single instance of each service would easily be able to handle quite a lot of load anyway?

**Comment by vjeux** [comment]
Right now we have a single instance for both, but they both are really far away from max utilization. The design of both were with scalability in mind. For json, it's just uploading / serving static files. For room, it's just moving opaque binary opaque across all the people in a single room, it doesn't do anything else and doesn't keep any state outside of the list of people in the room.

**Comment by kbariotis** [comment]
Amazing everyone, @NMinhNguyen thank you for containerizing `excalidraw-room`. [Here's the Docker image](https://hub.docker.com/repository/docker/excalidraw/excalidraw-room).

Now I am wrong to believe that the client is not configurable to use a specific URL for that service right? So what we can do now is make it configurable somehow so when someone builds the `excalidraw` image can pass the URL of the self-hosted `excalidraw-room` instance.

**Comment by NMinhNguyen** [comment]
> Now I am wrong to believe that the client is not configurable to use a specific URL for that service right? So what we can do now is make it configurable somehow so when someone builds the `excalidraw` image can pass the URL of the self-hosted `excalidraw-room` instance.

So they can do that now already as long as they configure this environment variable when building the image https://github.com/excalidraw/excalidraw/blob/54f8d8f820c064c5bd92b9058575719d68baca22/.env#L4

This isn't ideal because it requires a rebuild. However, I've been working on a solution that doesn't require rebuilds.

**Comment by dwelle** [comment]
We'll make it configurable for solutions like the `excalidraw-embed` React compo, but I have no experience with configuring docker images, apart from ENV vars like @NMinhNguyen said.

**Comment by kbariotis** [comment]
Yeah was just coming to say that! It's actually ready, we just need to put some documentation in place and next step is to do the same with storage server! Amazing!

**Comment by NMinhNguyen** [comment]
Storage server might be tricky due to its dependency on Google Cloud, so I rewrote it to rely on S3-compatible storage instead: https://github.com/NMinhNguyen/excalidraw-json

**Comment by kbariotis** [comment]
Yes indeed as it seems like it's using NDB which I'm not familiar at all. It's storage engine for mysql? Does it have the same API?

@vjeux @dwelle you were talking about migrating from GCP to Heroku. Will that include changing the data store as well? I think it's a great opportunity to choose a more well-known store so we can serve the self-hosting needs as well.

**Comment by dwelle** [comment]
Yes. I imagine we'd migrate to something like Heroku Postgres.

**Comment by vjeux** [comment]
Btw, we don't really need a database. We could use the file system.

**Comment by dwelle** [comment]
Probably not if this is accurate: 

> You don't want to be uploading anything to Heroku, it has an ephemeral file system which is reset on restarts/deploys. Anything uploaded should be uploaded to a permanent file store like Amazon S3 ---[source](https://stackoverflow.com/a/17509735/927631)

Which it probably is because I can't find anywhere on their plans what the disk size is.

**Comment by kbariotis** [comment]
If S3 and other similar services are our options, I would argue that is still something that can't be self-hosted but I don't have  strong opinion. 

I'm sure we can think of how it's best for hosting our own instance and then maybe rearchitect in such a way that would allow others to pop in other stores? E.g. have different DB adapters

Make sense?

**Comment by NMinhNguyen** [comment]
> If S3 and other similar services are our options, I would argue that is still something that can't be self-hosted but I don't have  strong opinion. 
> 
> I'm sure we can think of how it's best for hosting our own instance and then maybe rearchitect in such a way that would allow others to pop in other stores? E.g. have different DB adapters
> 
> Make sense?

Well, for S3 you have e.g. https://min.io/ which you can host locally with ease.

**Comment by kbariotis** [comment]
@NMinhNguyen I'm sure there are ton of similar services like that but that's not my point. The point is to keep it as simple as possible and there is nothing simpler than just supporting a MySQL where people can just point to a Docker image and get done with it. Also, stores like these are more well known and can easily be adopted where S3 clones may not be such well known.

**Comment by kbariotis** [comment]
But I do think that just having an adapters pattern we could just start supporting any store anybody wants.

**Comment by NMinhNguyen** [comment]
> @NMinhNguyen I'm sure there are ton of similar services like that but that's not my point. The point is to keep it as simple as possible and there is nothing simpler than just supporting a MySQL where people can just point to a Docker image and get done with it. Also, stores like these are more well known and can easily be adopted where S3 clones may not be such well known.

Could it be a matter of documentation/recommendation then? Because with MinIO you also "point to a Docker image and get done with it". I'm just wary of the extra overhead required to write these adaptors, but in theory not a bad idea.

**Comment by kbariotis** [comment]
Ok I see maybe I'm not aware of it, but really I don't have strong opinion on what people would prefer.

**Comment by bmullan** [comment]
Could you utilize something like IPFS ?

https://docs.ipfs.io/

Brian

**Comment by Turakar** [comment]
> > @NMinhNguyen, from my observations, the by far most common way of configuration for docker images is by environment variables. Sometimes there are also some command line parameters or config files involved, but not a manual build.
> 
> Essentially the problem here is that environment variables are read only once when you do `yarn build`. Unless you want to muck around with rewriting the build output, it’s pretty difficult to inject those variables. Further to that, most solutions I’ve seen will suffer from content hashing issues. A rebuild is simply the most robust way to inject those environment variables for a static website (CRA), but I’m open to suggestions.

Im sorry that the different points which need to be discussed are mixed in one thread, but I recently had some new ideas for the configuration issue after looking at riot-web (a common Matrix client). They use a config.json which is provided by the web server together with the application files as a static resource. The client parses this json. On the Docker front, the sysadmin mounts a custom config.json to the location of the default config.json for customization. Maybe you can prevent a manual rebuild this way.

**Comment by Jan-M** [comment]
+1 for moving config variables out of the build process and ideally towards environment variables for dependency URLs.

As for the technology of choice S3 or compatible API sound pretty good to me. Not a fan of just spin up "MySQL".

Ideally excalidraw API for GET/PUT/POST json would be well defined, so anyone could just implement their own.

In our deployment excalidraw sits behind oauth proxy, which would make the json storage api just some path `excalidraw..../json-store`

**Comment by nrgapple** [comment]
Hey just wanted to drop in and say this is awesome work! I created the very basic standalone version of excalidraw stripped of all public contact for our closed environment here at work. It is WAY behind the current version of excalidraw with all the amazing features that have been added in the last 6 months. 

I have been trying to follow this thread and the readme. So currently the docker image in the readme is stripped of the external server correct? And from this thread, is there a working solution to build a docker compose file of the client and the two backend parts (share server and collab sockets)? 

This docker image is awesome news to finally push an internalized version of excalidraw through our security.

**Comment by kbariotis** [comment]
Hi @nrgapple thanks for following this. At the moment, you can run this repo (the main UI client) along with the colab server in your own environment. We provide a Docker image for both and you point the UI to work with your own colab server.

The shared server is pending and has a little bit more work to do as we will have to migrate away from a Google's specific data storage to an open-source alternative.

Let us know if you wanna get involved in this. Thanks!

**Comment by dwelle** [comment]
> The shared server is pending and has a little bit more work to do as we will have to migrate away from a Google's specific data storage to an open-source alternative.

Where we (excalidraw.com) store data shouldn't be relevant as long as the server you point to adheres to a given protocol (API).

In the coming week I plan to lay ground to merge collab and storage servers. We can then clarify what API the server should conform to.

**Comment by nrgapple** [comment]
@kbariotis where would I point the UI to work with the colab server?

**Comment by NMinhNguyen** [comment]
> @kbariotis where would I point the UI to work with the colab server?

Wherever you deploy the collab server.

**Comment by nrgapple** [comment]
Sorry I guess what I mean is there an `.env` file or a place in the docker file where I point the the UI at the colab server

**Comment by nicksellen** [comment]
I hoped I'd be able to spin up a self-hosted instance of this, with local-only storage (local to the server that is, whether file system or local db), but seems this is not possible. 

> In the coming week I plan to lay ground to merge collab and storage servers

Does this mean merging the [socket server](https://github.com/excalidraw/excalidraw-room) and [storage server](https://github.com/excalidraw/excalidraw-json) into one codebase either nodejs or python? Where is the work at, I'm not sure I have capacity to contribute, but would be interested to know where the contribution point would be if I did!

The simplest with what I can see right now, would be to make [excalidraw-json](https://github.com/excalidraw/excalidraw-json) use a pluggable storage backend architecture, so you can use GAE, local filesystem, local postgres db, or whatever people want to implement... but if you're merging both projects, maybe that would be futile effort.

Thanks for the nice project in any case!

**Comment by dwelle** [comment]
Currently, we're slowly factoring out pretty much every IO (json server, collab WS server, localStorage,...) out of the core and exposing hooks that both we (excalidraw.com) and others can use to implement their own. After we're done, excalidraw.com will be using the very same API as if we were just another host app using the soon-to-be-published new version of the npm component.

At that point we'll probably document how best to implement your own server(s), but you could just as easily read the code to figure that out.

**Comment by nicksellen** [comment]
Ok, thanks for the update! I'll keep my eye out for the new versions... :+1:

**Comment by pandel** [comment]
Hi!

FWIW, I managed to set up a complete docker-compose stack to host excalidraw in our internal network, making use of @NMinhNguyen 's modification to ecalidraw-json and Minio S3 as backend datastore. It has no access to your Firebase database though, but collaboration and sharing is working fine, at least for us.

If you want to try it, you can clone this repo: https://gitlab.open-mind.space/pandel/excalidraw-docker

Perhaps it is useful for somebody :-)

Regards,
Holger

**Comment by DandyDev** [comment]
> Im sorry that the different points which need to be discussed are mixed in one thread, but I recently had some new ideas for the configuration issue after looking at riot-web (a common Matrix client). They use a config.json which is provided by the web server together with the application files as a static resource. The client parses this json. On the Docker front, the sysadmin mounts a custom config.json to the location of the default config.json for customization. Maybe you can prevent a manual rebuild this way.

I just wanted to chime in here. I think environment variables is still the most deployment-friendly way to configure things. Of course, these environment variables cannot be read at runtime by code running in the users browser. Having NGINX serve a config file that is mounted in the Docker container, is one way of solving this. A variation on that idea that goes back to the idea of environment variables, is to create a start-up script (shell) that basically wraps the command that starts NGINX, but before it runs NGINX, it will first read the environment variables that contain the necessary URLs and creates/modifies a `config.json` with that info.

So basically your `config.json` would be created at runtime, when the container starts, before the web server starts. No need to mount anything!

**Comment by FuckDoctors** [comment]
@pandel Thanks for your job, but it's `403 Forbidden`.
Can you share it on github?

**Comment by pandel** [comment]
@FuckDoctors Hey, has been a configuration error. Please try again, should work now :-)

**Comment by FuckDoctors** [comment]
@pandel It's working now. Thanks again.

**Comment by pandel** [comment]
@FuckDoctors Cool! If you have time, tell me what you think about it :-)

**Comment by MartinP7r** [comment]
@pandel thanks for providing that docker package. I've been trying to get it working on our local server but no luck so far. 
You've disabled issues on your gitlab repo. Is there a way to contact you for some advice?

**Comment by pandel** [comment]
@MartinP7r You can contact me via mail, see my Github profile :-).

**Comment by dpthegrey** [comment]
> So I'm opening the self-hosting thread.
> 
> ## Current architecture
> First an architecture overview. Excalidraw is made of three main parts at the moment.
> 
> * The client, what you see when you visit `excalidraw.com` which is this repo
> * The sockets server, which powers the collaboration feature, [it's this repo](https://github.com/excalidraw/excalidraw-room), a Node.js server running `socket.io`
> * The store server, which powers the sharing feature, [it's this repo](https://github.com/excalidraw/excalidraw-json), a Python server running on GAE with some kind of a DB
> 
> At the moment, we are planning on publishing a Docker image of the Excalidraw client only. That means that people will be able to self-host the client but they won't be able to use sharing/collaboration features because the client will still have to talk to our cloud-hosted backend servers on which there are CORS restrictions.
> 
> ## For now
> For the image that we are planning on publishing soon (in the next few days), we have already stripped out the google tag manager but I see the visible sharing/collaboration icons as an issue. People will be frustrated by not understanding why they don't work.
> 
> I suggest we strip them out too on the Docker image only. Temporarily until we have a plan for the rest two services. It's the best UX we can offer IMO.
> 
> ## For the future
> Moving forward we would like people to be able to self-host a full-fledged Excalidraw environment. My suggestion is to publish a Docker image for each of the two backend servers so people can self-host these as well. Having separate images promotes better scalability and also follows the one service/per container practice. They will be easier to maintain and also people can opt-out of a feature if they don't want to support, e.g. they only want collaboration features but not the storing feature.
> 
> Then we will be able to point the client image to specific URLs and it will simply work.
> 
> For that to work, the socket server will be much simpler as it has no dependencies. On the other hand, the storing server has a DB dependency which we will have to consider. Ideally, it should be something easy to use and maintain, like a postgres or a redis instance.
> 
> We could consider how the future of that service looks like, but for now, we can focus on what it actually needs.
> 
> Let me know your thoughts and we can start planning right away.
> 
> _I should also mention the [embedded Excalidraw instance](https://github.com/excalidraw/excalidraw-embed) we are working on that would also benefit from being able to connect to a self-host backend server but also have a better UX without icons that don't work._
> 
> ## Todo
> * [x]  Dockerize [main frontend client](https://hub.docker.com/r/excalidraw/excalidraw)
> * [x]  Dockerize [colab server](https://hub.docker.com/r/excalidraw/excalidraw-room)
> * [ ]  Dockerize storage server

How do we selfhost by combining Dockerized main frontend and Colab server together in a docker-compose file and self host it?

**Comment by KuenzelIT** [comment]
Hey guys,

did you make any progress in factoring out the colab and storage server. In my test just now the client still used the cloud colab and storage servers.

Thank you for working on a self hosted version!

**Comment by Turakar** [comment]
I am not from the team, but this might be interesting to you: https://github.com/excalidraw/excalidraw/issues/2213#issuecomment-754943207

**Comment by srkunze** [comment]
Is there an update on the documentation side of things? Or an API spec that needs to be fulfilled when implementing the storage server by oneself?

**Comment by Utopiah** [comment]
Hi, :wave: came here from https://github.com/excalidraw/excalidraw#self-hosting as I'm also interested in the full self-hosted solution, frontend with collaboration capabilities.

**Comment by vincentfretin** [comment]
@Utopiah I'm currently trying to host it myself. I use excalidraw-docker from @pandel  that uses excalidraw-json from @NMinhNguyen see https://github.com/excalidraw/excalidraw/issues/1772#issuecomment-738867944 Thank you both!
It works great with the excalidraw version from 1st march, commit 91eb8834e85caebeb795a20ae86d11ea091a753d
but the share whiteboard link that store the whiteboard to the bucket in minio doesn't work anymore with latest master, it tries to connect to firebase, so there probably some changes since then that broke it. A version between march and may 18 may still works, I didn't test that yet.
from the excalidraw-docker repo:
```
./build.sh
cd src_repo/excalidraw
git checkout 91eb8834e85caebeb795a20ae86d11ea091a753d
cd -
./build.sh --rebuild-all
```

**Comment by srkunze** [comment]
@Utopiah @vincentfretin 

I also got https://gitlab.open-mind.space/pandel/excalidraw-docker working as it was published.

Something weird happened regarding docker+nginx+https-redirect in Chrome. So I unmapped port 80 in `docker-compose.yml`.

**Comment by vincentfretin** [comment]
I was wrong in my last comment. The share link feature works on master, I had previously an error with Internal Server Error because the bucket wasn't created properly in minio. I'm not sure if it was an issue with `build.sh`, but I executed again the commands to create the bucket and it was ok.

With the version from march and current master, you get errors related to firebase if you have an empty config for firebase. To disable firebase completely, you can edit `src/excalidraw-app/data/firebase.ts` and comment all code in `saveToFirebase` and just `return true;` there, same for `loadFromFirebase` comment the code and just `return null;`.
From my understanding of the code, `loadFromFirebase` is used when you access an url with room param, `saveToFirebase` is triggered when you stop the collab session and each time it broadcasts the full scene elements every 20s if there were changes.

I didn't know what kind of database was firebase before I looked at the code in `firebase.ts`. The api reminds me of [kinto](https://github.com/Kinto/kinto) I used in a project, it uses a postgres db with the concepts of buckets, collections and entries. One could add a postgres and kinto containers to the `docker-compose.yaml` and modify `firebase.ts` to use [kinto-http.js](https://github.com/Kinto/kinto-http.js) to load and save from kinto.

**Comment by pandel** [comment]
> I was wrong in my last comment. The share link feature works on master, I had previously an error with Internal Server Error because the bucket wasn't created properly in minio. I'm not sure if it was an issue with `build.sh`, but I executed again the commands to create the bucket and it was ok.

Sometimes it seems as if the minio container needs a bit more time to start up got the first time. So I added a 5 second timeout to `build.sh` before I send the "make bucket" command to the container. Perhaps 5 seconds wasn't enough on your system. Glad to here that it worked on second try :-)

**Comment by KuenzelIT** [comment]
> @Utopiah @vincentfretin
> 
> I also got https://gitlab.open-mind.space/pandel/excalidraw-docker working as it was published.
> 
> Something weird happened regarding docker+nginx+https-redirect in Chrome. So I unmapped port 80 in `docker-compose.yml`. The problem were “too many redirects”.

My browser is unable to connect to that address.

**Comment by pandel** [comment]
> > @Utopiah @vincentfretin
> > I also got https://gitlab.open-mind.space/pandel/excalidraw-docker working as it was published.
> > Something weird happened regarding docker+nginx+https-redirect in Chrome. So I unmapped port 80 in `docker-compose.yml`. The problem were “too many redirects”.
> 
> My browser is unable to connect to that address.

That's correct. As I am moving to a new home at the moment, my servers are offline. I expect them to be online in a few days, when I was able to setup my infrastructure again. Sorry...

**Comment by AyushmanChatterjee** [comment]
> Hi!
> 
> FWIW, I managed to set up a complete docker-compose stack to host excalidraw in our internal network, making use of @NMinhNguyen 's modification to excalidraw-json and Minio S3 as backend datastore. It has no access to your Firebase database though, but collaboration and sharing is working fine, at least for us.
> 
> If you want to try it, you can clone this repo: https://gitlab.open-mind.space/pandel/excalidraw-docker
> 
> Perhaps it is useful for somebody :-)
> 
> Regards,
> Holger

Hi @pandel ,

Thanks a lot for sharing your repo with clear instructions. There were zero problems setting up. `build.sh` was super handy.

And much thanks to all all the amazing people who have helped build this project and worked towards dockerizing. Love Excalidraw!

**Comment by KuenzelIT** [comment]
> ~That's correct. As I am moving to a new home at the moment, my servers are offline. I expect them to be online in a few days, when I was able to setup my infrastructure again. Sorry...~
> 
> @KuenzelIT
> I am online again :-), so it should work now...

Hosting your servers at home? Nice, decentralization at its best. Thank you!

**Comment by KuenzelIT** [comment]
Are there any news on dockerizing the storage server?

**Comment by KuenzelIT** [comment]
Can we get a short info, whether this is still on the roadmap? We want to self host Excalidraw in our company, that is why we need an official way to self host it.

**Comment by vjeux** [comment]
There are no concrete plans in the roadmap of the current maintainers to streamline this. But this is something that we want to see happen and would accept and reviews PRs that make it happen.

**Comment by ruiaraujo** [comment]
@vjeux So PRs decouple the collaboration feature from Firebase (even supporting disabling it as an initial step) will be accepted, right?

**Comment by dwelle** [comment]
@ruiaraujo actually, no. Whatever is in `src/excalidraw-app` is just an implementation of https://excalidraw.com, and it's not planned to be exposed as an API — it will always contain ad-hoc code specific to the website. You are encouraged to check the source code to see how we're doing it so they can implement their own solution, or you can copy/paste it if you so desire.

That said, what _is_ on the roadmap is to factor some of it out into packages that we _will_ expose or publish on npm. But since it's closely dependent on our use cases, I'd say it's not open to contribution right now, at least not until we specify what exactly needs to be done and in what format.

**Comment by Kiliandeca** [comment]
Hi, if someone else is interested I recently worked on self-hosting Excalidraw as a side project and created 2 repositories and Docker images:

- [(GitLab) kiliandeca/excalidraw-fork](https://gitlab.com/kiliandeca/excalidraw-fork): A fork of Excalidraw react app which supports configuration as runtime environment variable meaning you can use the docker image and change your config without rebuilding the project. I also made some changes to support using an alternative HTTP backend storage server instead of Firebase to store images and collab scenes.
- [(GitLab) kiliandeca/excalidraw-storage-backend](https://gitlab.com/kiliandeca/excalidraw-storage-backend): Similar to excalidraw-json you can store exported scenes. You can use it with the database of your choice (in memory, redis, mongo, postgres and mysql). It can also be used with my front-end fork to replace Firebase for images and collab scenes.

You can use them with the official [excalidraw-room image](https://hub.docker.com/r/excalidraw/excalidraw-room), I made an example docker-compose file: [examples/docker-compose.yml](https://gitlab.com/kiliandeca/excalidraw-fork/-/blob/main/examples/docker-compose.yml)

**Comment by christf** [comment]
> Hi, if someone else is interested I recently worked on self-hosting Excalidraw as a side project and created 2 repositories and Docker images:
> 
>     * [(GitLab) kiliandeca/excalidraw-fork](https://gitlab.com/kiliandeca/excalidraw-fork): A fork of Excalidraw react app which supports configuration as runtime environment variable meaning you can use the docker image and change your config without rebuilding the project. I also made some changes to support using an alternative HTTP backend storage server instead of Firebase to store images and collab scenes.
> 
>     * [(GitLab) kiliandeca/excalidraw-storage-backend](https://gitlab.com/kiliandeca/excalidraw-storage-backend): Similar to excalidraw-json you can store exported scenes. You can use it with the database of your choice (in memory, redis, mongo, postgres and mysql). It can also be used with my front-end fork to replace Firebase for images and collab scenes.
> 
> 
> You can use them with the official [excalidraw-room image](https://hub.docker.com/r/excalidraw/excalidraw-room), I made an example docker-compose file: [examples/docker-compose.yml](https://gitlab.com/kiliandeca/excalidraw-fork/-/blob/main/examples/docker-compose.yml)

Very nice. What would be required to get that merged upstream?

**Comment by JannikStreek** [comment]
@christf @Kiliandeca FYI: I built on top of Kilians work and merged with upstream. Feel free to test: https://github.com/b310-digital/excalidraw
I also changed (optimized) the docker setup a bit to my needs.

**Comment by christf** [comment]
hm. I am on debian and docker-compose tells me:

ERROR: Version in "./docker-compose.yml" is unsupported. You might be seeing this error because you're using the wrong Compose file version. Either specify a supported version (e.g "2.2" or "3.3") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1.
I had to remove the target line after changing the version to 3.3. I am not sure if I can get it to work that way.

**Comment by christf** [comment]
... so I guess I am running into the "collaboration does not work in some circumanstence (no tls and non-localhost)"

```
Unhandled Rejection (TypeError): window.crypto.subtle is undefined
generateEncryptionKey
src/data/encryption.ts:16

  13 | >(
  14 |   returnAs?: T,
  15 | ): Promise<T extends "cryptoKey" ? CryptoKey : string> => {
> 16 |   const key = await window.crypto.subtle.generateKey(
  17 |     {
  18 |       name: "AES-GCM",
  19 |       length: ENCRYPTION_KEY_BITS,

generateCollaborationLinkData
src/excalidraw-app/data/index.ts:150

  147 | 
  148 | export const generateCollaborationLinkData = async () => {
  149 |   const roomId = await generateRoomId();
> 150 |   const roomKey = await generateEncryptionKey();
  151 | 
  152 |   if (!roomKey) {
  153 |     throw new Error("Couldn't generate room key");

This screen is visible only in development. It will not appear if the app crashes in production.
Open your browser’s developer console to further inspect this error.  Click the 'X' or hit ESC to dismiss this message.
```

**Comment by christf** [comment]
... so how do I use the collaboration feature with this when terminating tls in front of excalidraw?

**Comment by christf** [comment]
> @christf @Kiliandeca FYI: I built on top of Kilians work and merged with upstream. Feel free to test: https://github.com/b310-digital/excalidraw I also changed (optimized) the docker setup a bit to my needs.

I got it to deploy and can use it. The only thing I am still struggling with is the collaboration feature (see above). Thank you for your work!

**Comment by JannikStreek** [comment]
> > @christf @Kiliandeca FYI: I built on top of Kilians work and merged with upstream. Feel free to test: https://github.com/b310-digital/excalidraw I also changed (optimized) the docker setup a bit to my needs.
> 
> I got it to deploy and can use it. The only thing I am still struggling with is the collaboration feature (see above). Thank you for your work!

you mean the problem with `window.crypto.subtle is undefined` ? I got this when trying to access it locally via ip, but using localhost it works. If hosted on your server you would need an https connection as the feature is only supported within a secure context. I can't really tell more, because I don't know what kind of setup you are using.

**Comment by christf** [comment]
Yes, that one. I am using the docker-cokpose-prod file on host a and access the service from host b. Maybe I just need a local reverse proxy in front of it?

**Comment by matbgn** [comment]
@JannikStreek I see many commits behind the stream 
> This branch is [4 commits ahead](https://github.com/b310-digital/excalidraw/compare/excalidraw:excalidraw:master...master), [31 commits behind](https://github.com/b310-digital/excalidraw/compare/master...excalidraw:excalidraw:master) excalidraw:master.

What's your plan exactly ? Did you try a PR or plan it ? Do you plan to merge those commits in your fork ?

**Comment by alswl** [comment]
@JannikStreek Maybe using cherry-pick mode is easier to maintain than rebase mode? I see that the upstream CHANGELOG.md describes a lot of new content, and there are no relevant changes in your repository yet.

**Comment by alswl** [comment]
Update:

Hi, everybody, I provided self-host collaboration enabled excalidraw with docker-compose.
The repo is here, https://github.com/alswl/excalidraw-collaboration

Now, with the simple commands(`docker-compose up`), you can get a self-host and can share the room link to your friends.

**Comment by matbgn** [comment]
> Update:
> 
> Hi, everybody, I provided self-host collaboration enabled excalidraw with docker-compose. The repo is here, https://github.com/alswl/excalidraw-collaboration
> 
> Now, with the simple commands(`docker-compose up`), you can get a self-host and can share the room link to your friends.

Do you plan to more or less follow the main project or is it a one shot release?

**Comment by alswl** [comment]
@matbgn Actually, there are tow version released, v0.12.0 and v0.13.0 .
If the API of upstream did not have break changes, I will follow the upstream upgrades.

**Comment by matbgn** [comment]
Looks really promising! Could you publish it on Docker Hub or any OCI registry platform and tag its version? Or at the very least could you use Github's releases tag on the master excalidraw-collaboration repo?

**Comment by alswl** [comment]
@matbgn The solution required build localhost now, the packager should set some parameters before building image.
Maybe I could do more works to parametrize the `Dockerfile` and `docker-compose.yaml`.

**Comment by JannikStreek** [comment]
fyi: I have rebased my repository to include the latest commits (master): https://github.com/b310-digital/excalidraw 

@alswl i have squashed my changes to just one commit, should be easier now to maintain. So I don't really need cherry picking :)

**Comment by alswl** [comment]
@JannikStreek Jannik cool.

Now I provided a one-click to deploy solution: https://github.com/alswl/excalidraw-collaboration .
The repo includes forked excalidraw, room and storage.

More detail is described at https://en.blog.alswl.com/2022/10/self-hosted-excalidraw/

**Comment by perryrh0dan** [comment]
Any update on officially supporting this ?

**Comment by Digital39999** [comment]
Why are you referencing this if only ways of self hosting are months old and outdated, why is excalidraw so insisted on firebase and so self host unfriendly.

**Comment by janderssonse** [comment]
I work for an Agency, and as many other agencys using more and more Open Source, self hosting possibilties is an essential base demand for taking on new services. For example, Excalidraw is one project that fills a need I know many agencies have, but at it current state, this is an essential feature - easy, full self-hosting.. I think excalidraw would get much more leverage if they offered total selfhosting possibilities , but for more advanced features an pro-verison, like The GitLab way and others do. 

Looking at forks to solve this, it is nice inititative for some,  but for all means, as an organisation we would much rather have the saftey of using the original, official product, with updated patches etc. This Issue has been open for years,  and for excalidraw I can only see a win-win situation by offering this. Excalidraw will get more adopters and users in the long term, and organisations can fully use and expand upon

So my question to excalidraw is:

 - Are there any plans in the roadmap to offer this in the near future? (The most urgent one being this one - https://github.com/excalidraw/excalidraw/issues/4038 I think).
 - Would PR for https://github.com/excalidraw/excalidraw/issues/4038 be accepted?

**Comment by saghul** [comment]
We've integrated Excalidraw in Jitsi and tried to engage so we could discuss and upstream some changes we needed.

https://github.com/excalidraw/excalidraw/discussions/5902

Not one response alas.

Based on this issue I think there simply is no interest at all in decoupling from Firebase.

I am very thankful for Excalidraw, I just wish some of its features were easy to toggle on or off by embedders.

**Comment by HackerFoo** [comment]
@janderssonse @saghul  I had likewise been looking for a way to self-host Excalidraw, and I found that it was integrated into Jitsi Meet, and there is an [option in NixOS to enable it](https://search.nixos.org/options?channel=23.11&show=services.jitsi-meet.excalidraw.enable).

I've tested Excalidraw in Jitsi Meet and it seems to work well.

**Comment by Digital39999** [comment]
It is nice that it's integrated there, but that's still not a solution to fully self-host excalidraw.

**Comment by saghul** [comment]
I didn't claim it was. My point is that we also tried to collaborate because we like and use it, but got no response.

**Comment by manumitted** [comment]
I haven't looked into this closely, so please bear with me, but my team is looking for a self-hosted whiteboard tool to work with Nextcloud.  It seems there's a lot of discussion above about storage, and I just want to use Nextcloud as our storage and user management system anyway -- I just want to add the collaborative whiteboard functionality.  Maybe the way Collabora works -- it has its own server, but the data files always at least appear to be hosted on Nextcloud.

**Comment by Digital39999** [comment]
This whole issue would be lot simpler if maintainers wouldn't ignore this, it's not a big deal switching firestore for supabase and removing local storing, many have done it. But they are all outdated and noone wants to use non official code, but this is getting ignored for last quite few years and for unknown reasons it'll likely never happen that we can actually selfhost this in its fully glory.

**Comment by manumitted** [comment]
It would be ideal to have direct access to the latest, but all things considered, I'm thinking I might start looking into the project forks.

**Comment by lietu** [comment]
So looking into this it seemed running this depends on excalidraw, excalidraw-room, and excalidraw-store. I saw the [excalidraw-store](https://github.com/excalidraw/excalidraw-store) was set up with some Google Cloud Storage integration, I didn't know what it was for but I thought it was probably necessary for this whole thing to work locally I quickly built [excalidraw-alternative-store](https://github.com/cocreators-ee/excalidraw-alternative-store) to just have a quick local filesystem storage and be extensible for some better implementation future.

Then I tried to whip up a [`docker-compose.yml`](https://github.com/cocreators-ee/excalidraw-alternative-store/blob/main/docker-compose.tpl.yml) since this seemed like a problem that should be easy to solve, turns out it's at least rather confusing.

What I've set up with it is:

 - Easily accessible "backend" Docker image with excalidraw-alternative-store with zero security, the "protocol" designed for it is .. depressingly simplistic
 - [excalidraw-room](https://github.com/excalidraw/excalidraw-room) for realtime collaboration (looks like just a socket.io server?)
 - Custom [excalidraw](https://github.com/cocreators-ee/excalidraw-alternative-store/blob/main/excalidraw/Dockerfile) Docker image that allows passing the necessary environment variables for configuration
 
 .. unfortunately due to the way excalidraw's `yarn build:app:docker` -task builds to static files, it's necessary to run it at container startup time instead of during build time. It's sub-optimal, but works, at least if your docker host has enough memory allocated to it and you can wait the little while for it to start up.
 
Tested the `docker-compose` setup, put the containers behind Nginx Proxy Manager (enabling WebSockets support for at least the "room"), to get the components nice https:// addresses and configured the CORS variables appropriately, and it runs, but:

- It is making some random requests to `firestore.googleapis.com` and `firebasestorage.googleapis.com`, didn't yet try to look at what this is for
- It also tried to make some requests to `https://scripts.simpleanalyticscdn.com/latest.js` but gladly it was blocked by something
- Menu -> Save to ... -> Excalidraw+ sends me to some random place

However what works is

- Self-hosted excalidraw frontend overall is available
- Menu -> Save to ... -> Export to Link submits the contents to `excalidraw-alternative-store` and gives me a working link
- Real-time collaboration seems to work via my self-hosted excalidraw-room

To get this thing to be properly self-hosted, it seems whatever is making those firestore/firebase requests should be removed/disabled/replaced, the analytics should be disabled, and that "Excalidraw+" thing needs to either be replaced or disabled.

**Comment by Digital39999** [comment]
@lietu as far as I know, Excalidraw has mixed saving ways where one uses json store, and the other one uses firestore which is most likely for collaboration rooms, best way would be if this repository migrated to supabase and dropped json store support, that seems like ideal solution.

**Comment by Someone0nEarth** [comment]
I was looking for a _fully_ self hosted excalidraw solution.. many were still using Firebase for the storage.
I used [this](https://github.com/alswl/excalidraw-collaboration) as an inspiration for a fully self hosted solution using traefik and mongodb. Export to Link and collaboration are working as fine.

https://github.com/Someone0nEarth/excalidraw-self-hosted

**Comment by PatWie** [comment]
Seeing also the need for self-hosting, I did start a self-containing Golang implementation providing collaboration and storage options (S3, SQLite, filesystem, in-memory)
https://github.com/PatWie/excalidraw-complete

I simply don't see any reason for pasting together several docker images, when I can simply put a single binary somewhere and can even cross-compile it for Mac.

While it works with the official excalidraw UI without patching - besides setting endpoints, I am also struggling with the firebase dependency.

**Comment by Someone0nEarth** [comment]
> Seeing also the need for self-hosting, I did start a self-containing Golang implementation providing collaboration and storage options (S3, SQLite, filesystem, in-memory) https://github.com/PatWie/excalidraw-complete

I like your attempt!
 
> While it works with the official excalidraw UI without patching - besides setting endpoints, I am also struggling with the firebase dependency.

At the moment, there are only two ways to get rid of the Google Firebase/-storage usage when using a self-hosted Excalidraw version: 

    1. Using a patched version of Excalidraw, which is supporting a different backend storage solution (like [this](https://github.com/alswl/excalidraw)) in combination with a self-hosted backend storage service like [this](https://github.com/alswl/excalidraw-storage-backend)
    2. Implementing/using a self-hosted version of Firestorage (API) to use it with an unpatched version of Excalidraw.

I would recommend to update the readme of your project, that in the end, your self-hosted version uses Googles Firestorage (in collaboration mode or when "Save to..."->"Export to Link" is used). 

A way to test, that Googles Firebase storage is still used, can be found [here](https://github.com/Nenodema/excalidraw-self-hosted-stack/issues/3#issue-2154065668).

**Comment by PatWie** [comment]
Wow! Thanks for that information! I was unaware of that.

Is dislike patching many places in the front-end code (I am not a front-end guy). As the go binary serves the static files anyway, one can patch those calls in the assets on-the-fly and provide an API stub that somehow covers the functions needed for those calls.

Here is a prototype that has both API endpoints (documents:commit, documents:batchGet)
https://github.com/PatWie/excalidraw-complete/tree/firebase-patch

It removes most of the error messages but is merely a proof-of-concept in the current stage. I havent fully grasped yet that `https://firestore.googleapis.com/google.firestore.v1.Firestore/Listen/channel?database=projects....` calls and I definitely not gonna re-implement firebase.


> A way to test, that Googles Firebase storage is still used, can be found https://github.com/Nenodema/excalidraw-self-hosted-stack/issues/3#issue-2154065668.

[This branch](https://github.com/PatWie/excalidraw-complete/tree/firebase-patch) has no longer the issue.

**Comment by filwu8** [comment]
![1714482963673](https://github.com/excalidraw/excalidraw/assets/54697960/7d221f15-3438-4e12-af8f-82d18079e855)
why？

https://firestore.googleapis.com/v1/projects/excalidraw-room-persistence/databases/(default)/documents:batchGet

**Comment by andraskuti** [comment]
I have created the following diagram to better understand the architecture and dependencies of self-hosted Excalidraw primarily from a feature perspective (saving, link sharing, collab) based on my current understanding.

What we see here is the vanilla case, where all components are from github.com/excalidraw/* repos.  
Obviously custom solutions could change any part of the system.
Library and AI features and Excalidraw+ is not part of this diagram.

Hope this helps others getting started. Any suggestion is welcome. 

<img width="1122" alt="image" src="https://github.com/excalidraw/excalidraw/assets/1612936/a8df6b85-76a4-4616-a5d3-a748278ac15d">

**Comment by AlphaCraft9658** [comment]
The official docker image has not been updated in a very long while.

**Comment by ad1992** [comment]
Hi PR is welcome to update docker image

**Comment by AlphaCraft9658** [comment]
> Hi PR is welcome to update docker image

Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?

**Comment by ad1992** [comment]
> > Hi PR is welcome to update docker image
> 
> Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?

Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it

**Comment by AlphaCraft9658** [comment]
> > > Hi PR is welcome to update docker image
> > 
> > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> 
> Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it

You mean in the GitHub actions? I don't really know how they work, but it seems like there is some server error when fetching the packages for the build. I'll try to fork and look into it.

**Comment by ad1992** [comment]
@AlphaCraft9658 you should be able to run those commands in local and test it out as well `in publish-docker.yml`

**Comment by AlphaCraft9658** [comment]
> @AlphaCraft9658 you should be able to run those commands in local and test it out as well in `publish-docker.yml`

I got it. Now I'll see what's going on. I added a workflow_dispatch trigger for the workflow, for testing.

**Comment by AlphaCraft9658** [comment]
> > > Hi PR is welcome to update docker image
> > 
> > 
> > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> 
> Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it

The docker build workflow is still failing. That's the current status: https://github.com/AlphaCraft9658/excalidraw/actions/runs/8970588981/job/24634460317

It cannot resolve index.html.

**Comment by AlphaCraft9658** [comment]
> > > Hi PR is welcome to update docker image
> > 
> > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> 
> Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it

How is it even supposed to be build? I see it's trying to use vite and cross-env. Something makes vite unable to resolve the index.html file.

**Comment by Hufschmidt** [comment]
> > > > Hi PR is welcome to update docker image
> > > 
> > > 
> > > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> > 
> > 
> > Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it
> 
> How is it even supposed to be build? I see it's trying to use vite and cross-env. Something makes vite unable to resolve the index.html file.

There are various open PRs to fix the docker build, non have been merged so far (for unknown reasons).
See:
- https://github.com/excalidraw/excalidraw/pull/7806
- https://github.com/excalidraw/excalidraw/pull/7502
- https://github.com/excalidraw/excalidraw/pull/7430
- https://github.com/excalidraw/excalidraw/pull/5381

Personally I've since moved on to [TlDraw](https://github.com/tldraw/tldraw) which seems to have implemented a real open-source collaboration workflow unlike with ExcaliDraw, where I see 290 open pull requests, many going back as fas as 2020. (As is also this issue)

**Comment by AlphaCraft9658** [comment]
> > > > > Hi PR is welcome to update docker image
> > > > 
> > > > 
> > > > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> > > 
> > > 
> > > Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it
> > 
> > How is it even supposed to be build? I see it's trying to use vite and cross-env. Something makes vite unable to resolve the index.html file.
> 
> There are various open PRs to fix the docker build, non have been merged so far (for unknown reasons).
> See:
> - https://github.com/excalidraw/excalidraw/pull/7806
> - https://github.com/excalidraw/excalidraw/pull/7502
> - https://github.com/excalidraw/excalidraw/pull/7430
> - https://github.com/excalidraw/excalidraw/pull/5381
> 
> Related Issues:
> - https://github.com/excalidraw/excalidraw/issues/7582
> - https://github.com/excalidraw/excalidraw/issues/7403
> - https://github.com/excalidraw/excalidraw/issues/6839
> - https://github.com/excalidraw/excalidraw/issues/5163
> - https://github.com/excalidraw/excalidraw/issues/5066
> 
> `<rant>`
> Personally I've since moved on to [TlDraw](https://github.com/tldraw/tldraw) which seems to have managed to implemented a working open-source collaboration workflow. Very much unlike ExcaliDraw, where I can see as many as 290 open pull requests , many going back as fas as 2002, as is also this issue by the way. All despite there beeing new releases/tags every now and then...
> `</rant>`

TLDraw doesn't have an up-to-date docker image either.

**Comment by ad1992** [comment]
> > > > Hi PR is welcome to update docker image
> > > 
> > > 
> > > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> > 
> > 
> > Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it
> 
> How is it even supposed to be build? I see it's trying to use vite and cross-env. Something makes vite unable to resolve the index.html file.

The `index.html` should be available in the `public` folder so you will need to update the path correctly

**Comment by AlphaCraft9658** [comment]
> > > > > Hi PR is welcome to update docker image
> > > > 
> > > > 
> > > > Is there any information on how the docker image is built? I tried setting it up myself, but haven't been able to get it working. Is it the dockerfile and compose file that's already available in the repo?
> > > 
> > > 
> > > Yes the files are already available in the repo, the build is not. being published as its failing so you will need to check why the publish docker gh action is failing and fix it
> > 
> > How is it even supposed to be build? I see it's trying to use vite and cross-env. Something makes vite unable to resolve the index.html file.
> 
> The `index.html` should be available in the `public` folder so you will need to update the path correctly

There is also an index.html in excalidraw-app. And where would I set this path? There are all kinds of different configs.

**Comment by ad1992** [comment]
@AlphaCraft9658 yes you are right, it should be `index.html` inside the `excalidraw-app` folder - thats the main entry file responsible for [excalidraw.com](https://excalidraw.com)

**Comment by AlphaCraft9658** [comment]
> @AlphaCraft9658 yes you are right, it should be `index.html` inside the `excalidraw-app` folder - thats the main entry file responsible for [excalidraw.com](https://excalidraw.com)

But do you know how to specify that path? I haven't worked with vite ever. The error happens during the actual build process with vite, which fails due to the "missing" index.html file.

**Comment by ad1992** [comment]
> > @AlphaCraft9658 yes you are right, it should be `index.html` inside the `excalidraw-app` folder - thats the main entry file responsible for [excalidraw.com](https://excalidraw.com)
> 
> But do you know how to specify that path? I haven't worked with vite ever. The error happens during the actual build process with vite, which fails due to the "missing" index.html file.

can you update the command to 👇🏻 and try
```
yarn --cwd ./excalidraw-app build
```

**Comment by AlphaCraft9658** [comment]
> > > @AlphaCraft9658 yes you are right, it should be `index.html` inside the `excalidraw-app` folder - thats the main entry file responsible for [excalidraw.com](https://excalidraw.com)
> > 
> > But do you know how to specify that path? I haven't worked with vite ever. The error happens during the actual build process with vite, which fails due to the "missing" index.html file.
> 
> can you update the command to 👇🏻 and try
> ```
> yarn --cwd ./excalidraw-app build
> ```

That causes a loop, as it seems. But we're talking about another PR which may solve it, you know.

**Comment by AlphaCraft9658** [comment]
> > > @AlphaCraft9658 yes you are right, it should be `index.html` inside the `excalidraw-app` folder - thats the main entry file responsible for [excalidraw.com](https://excalidraw.com)
> > 
> > But do you know how to specify that path? I haven't worked with vite ever. The error happens during the actual build process with vite, which fails due to the "missing" index.html file.
> 
> can you update the command to 👇🏻 and try
> ```
> yarn --cwd ./excalidraw-app build
> ```

That causes a loop, as it seems. But we're talking about another PR which may solve it, you know.

**Comment by kiranatious** [comment]
Hi All,
if someone is still looking for a solution to use Excalidraw as a standalone.. its possible.
we need to run Excalidraw inside a docker nginx
2. excalidraw json (for websocket room)
3. firebase account for saving files and scenes

**Comment by saghul** [comment]
The concept of standalone is at odds with the Firebase requirement, IMHO.

**Comment by Digital39999** [comment]
Honestly its easier to make an app that wraps excalidraw package and selfhost modified excalidraw json for own custom storage.

**Comment by hex-m** [comment]
[Nextcloud Whiteboard](https://github.com/nextcloud/whiteboard) is based on Excalidraw and should be relatively easy to self-host.

**Comment by krismarc** [comment]
> Hey, just chiming in to get a response from the maintainers confirming whether or not this is on the roadmap to support (if this is even actionable), and if so, is there an ETA based on the current workload. Thanks.

well, I believe it's their business model and how they actually make money. No offense to the project, just saying 😄

**Comment by GetPsyched** [comment]
Yeah, that's fine. I was just looking for a response -- even if it's just confirmation that they won't officially support this.

To your point, a lot of projects to allow (and officially support) self hosting even if their business model is to provide hassle-free hosting. An example is Weblate -- see "Gratis" tier in https://weblate.org/en/hosting/

**Comment by Anynomouss** [comment]
Also looking forward to a full fledged self hosting solution. Could work with donations. Most people will chose to pay for the convenience of having it hosted for them, in the end it will only attract more users.

**Comment by katzeprior** [comment]
On the road it lists `Self-host Excalidraw+ on your own premises.` can I get some more information about this, I love excalidraw but would love to self-host the + version on prem.

**Comment by svefro** [comment]
+1 on the storage server.

**Comment by WhyNotHugo** [comment]
As far as I can tell (please do correct me if I'm missing something):

[excalidraw-room](https://github.com/excalidraw/excalidraw-room) can be used as a websocket server.

Firebase is the only available approach for a storage server and is required for collaboration. Does this implies that collaboration is entirely unfeasible for self-hosted instances?

**Comment by WhyNotHugo** [comment]
If I follow the logic correctly, clients just submit an encrypted payload into the server. If there's a version conflict, the server returns its current version, otherwise, it retains the version submitted by the client as latest.

Writing a tiny replacement in Go should be pretty straightforward, but I'm concerned that actually using it would require pretty invasive changes to excalidraw.

**Comment by saghul** [comment]
You don't need Firebase for the collaboration, the WS server is enough.

**Comment by WhyNotHugo** [comment]
This doesn't match my experience, although I may be missing something else:

If I don't override `VITE_APP_FIREBASE_CONFIG`, when I initiate collaboration, I see a bunch of requests to firebase, both from the initiator and the joinee, and collaboration works.

If I compile with `export VITE_APP_FIREBASE_CONFIG='{}'`, live collaboration doesn't work; the joinee simply sees a blank canvas and an error accessing the backend.

Maybe there's not a strong dependency but the code paths are somehow connected in a way that live collaboration doesn't work with Firebase disabled?

**Comment by saghul** [comment]
Did you configure the collaboration backend?

If so, things might have changed lately (for the bad, apparently) then.

We embed Excalidraw in Jitsi Meet and removed the Firebase part and collaboration works with just the room backend.

**Comment by WhyNotHugo** [comment]
> Did you configure the collaboration backend?

My entire build process is:

```
yarn install --frozen-lockfile --ignore-engines

cd excalidraw-app
export NODE_OPTIONS="--max-old-space-size=4096"
export VITE_APP_DISABLE_SENTRY=true
export VITE_APP_FIREBASE_CONFIG='{}'
../node_modules/.bin/vite build --sourcemap=false
```

Have I missed something?

**Comment by saghul** [comment]
Collaboration requires a backend: https://github.com/excalidraw/excalidraw-room

Not sure how you can set it using the unmodified source, sorry.
